File: DevelopArchitecture/DevelopArchitecture/threeclass/JProgressHUD/JProgressHUD.m
//
//  JProgressHUD.m
//  DevelopArchitecture
//
//  Created by Jeffrey on 2017/11/17.
//  Copyright © 2017年 Jeffrey. All rights reserved.
//

#import "JProgressHUD.h"
#import "JConstantHelp.h"
#import "JAppViewExtend.h"
static JProgressHUD *progressHUD =nil;
@interface JProgressHUD()
@property(nonatomic,strong)UILabel *m_titleLabel;
@end
@implementation JProgressHUD{
    UIImageView *progressBgImageView,*progressImageView;
    
}
-(id)init{
    if(self =[super init]){
        self.frame = CGRectMake(0,0,KWidth,KHeight);
        self.backgroundColor = COLOR_CREAR;
        progressBgImageView =[UIImageView funj_getImageView:CGRectMake((KWidth-202/2.5)/2, (KHeight-202/2.5)/2, 202/2.5, 202/2.5) image:@"reloadProgress_center"];
        [self addSubview:progressBgImageView];
        progressImageView =[UIImageView funj_getImageView:CGRectMake((progressBgImageView.frame.size.width-188/2.5)/2, (progressBgImageView.frame.size.height-188/2.5)/2, 188/2.5, 188/2.5) image:@"reloadProgress_route"];
        [progressBgImageView addSubview:progressImageView];
        self.tag = NSNotFound-1;

        UITapGestureRecognizer *tapGest =[[UITapGestureRecognizer alloc]initWithTarget:self  action:@selector(funj_stopProgressAnimate)];
        [self addGestureRecognizer:tapGest];
    }
    return self;
}

-(UILabel*)m_titleLabel{
    if(!_m_titleLabel){
        _m_titleLabel =[UILabel funj_getLabel:CGRectMake(0, 0, 0, 30) :JTextFCMakeAlign(PUBLIC_FONT_SIZE17, COLOR_TEXT_BLACK, NSTextAlignmentCenter)];
        [self addSubview:_m_titleLabel];
    }
    return _m_titleLabel;
}
-(void)funj_addProgressAnimate{
    CABasicAnimation *anima = [CABasicAnimation animationWithKeyPath:@"opacity"];
    anima.fromValue = [NSNumber numberWithFloat:0.3f];
    anima.toValue = [NSNumber numberWithFloat:1.0f];
    anima.duration = 1.0f;
    [progressBgImageView.layer addAnimation:anima forKey:@"opacityAniamtion"];
    
    CABasicAnimation* rotationAnimation;
    rotationAnimation = [CABasicAnimation animationWithKeyPath:@"transform.rotation.z"];
    rotationAnimation.toValue = [NSNumber numberWithFloat: M_PI * 2.0 ];
    rotationAnimation.duration = 0.6;
    rotationAnimation.cumulative = YES;
    rotationAnimation.repeatCount = HUGE_VALF;
    [progressImageView.layer addAnimation:rotationAnimation forKey:@"rotationAnimation"];
}
-(void)funj_stopProgressAnimate{
    [progressBgImageView.layer removeAllAnimations];
    [progressImageView.layer removeAllAnimations];
    [self removeFromSuperview];
}
-(void)funj_showProgressView{
    UIView *supView =[JAppViewTools funj_getKeyWindow];
    UIView *upProgressView =[supView viewWithTag:NSNotFound -1];
    if([upProgressView isKindOfClass:[JProgressHUD class]]){
        [(JProgressHUD*)upProgressView funj_stopProgressAnimate];
    }
    self.frame = CGRectMake(0,0,KWidth,KHeight);
    progressBgImageView.frame = CGRectMake((KWidth-202/2.5)/2, (KHeight-202/2.5)/2, 202/2.5, 202/2.5);
    [supView addSubview:self];
    [self funj_addProgressAnimate];
}
-(void)funj_showProgressView:(NSString*)title{
    UIView *supView =[JAppViewTools funj_getKeyWindow];
    UIView *upProgressView =[supView viewWithTag:NSNotFound -1];
    if([upProgressView isKindOfClass:[JProgressHUD class]]){
        [(JProgressHUD*)upProgressView funj_stopProgressAnimate];
    }
    self.frame = CGRectMake(0,0,KWidth,KHeight);
    progressBgImageView.frame = CGRectMake((KWidth-202/2.5)/2, (KHeight-202/2.5)/2, 202/2.5, 202/2.5);
    self.m_titleLabel.frame = CGRectMake(progressBgImageView.frame.origin.x-20, progressBgImageView.frame.size.height+progressBgImageView.frame.origin.y, progressBgImageView.frame.size.width+40, 30);
    self.m_titleLabel.text = title;
    [supView addSubview:self];
    [self funj_addProgressAnimate];
}

@end

static JMProgressHUD *mprogressHUD =nil;

@interface JMProgressHUD()
@property(nonatomic,assign)MprogressType m_progressType;
@property(nonatomic,assign)CGFloat m_time;

@property(nonatomic,weak)UIView* m_superView;
@property(nonatomic,strong)UILabel* m_titleLabel;
@property(nonatomic,strong)NSTimer *m_timer;
@property(nonatomic,strong)UIActivityIndicatorView *m_activityView;
@property(nonatomic,strong)UIImageView *m_blackAlphaView;
@property(nonatomic,copy)completeCallback m_callback;
@property(nonatomic,assign)BOOL m_isHasFinishCallback;

@end
@implementation JMProgressHUD
+(id)share{
    if(!mprogressHUD){
        mprogressHUD=[[JMProgressHUD alloc]init];
        mprogressHUD.backgroundColor = COLOR_CREAR;
        [mprogressHUD funj_addblackView];
    }
    return mprogressHUD;
}
-(id)initWithView:(UIView*)superView t:(MprogressType)type{
    if(self =[super initWithFrame:CGRectZero]){
        [self funj_reloadSuperView:superView t:type];
        self.backgroundColor = COLOR_CREAR;
        [self funj_addblackView];
    }
    return self;
}
-(void)funj_addblackView{
    _m_blackAlphaView =[UIImageView funj_getBlackAlphaView:CGRectZero];
    [self addSubview:_m_blackAlphaView];
    [_m_blackAlphaView funj_setViewCornerLayer:JFilletMake(1, 10, COLOR_CREAR)];
}
-(void)funj_reloadSuperView:(UIView*)superView t:(MprogressType)type{
    _m_progressType = type;
    _m_superView = superView;
}
-(UILabel*)m_titleLabel{
    if(!_m_titleLabel){
        _m_titleLabel =[UILabel funj_getLabel:CGRectZero :JTextFCMakeAlign(PUBLIC_FONT_SIZE15, COLOR_WHITE,NSTextAlignmentCenter)];
        [self addSubview:_m_titleLabel];
        _m_titleLabel.numberOfLines = 2;
    }
    return  _m_titleLabel;
}
-(NSTimer*)m_timer{
    if(!_m_timer){
        _m_timer=[NSTimer scheduledTimerWithTimeInterval:2 target:self  selector:@selector(funj_stopProgressAnimate) userInfo:nil repeats:YES];
    }
    return _m_timer;
}
-(UIActivityIndicatorView*)m_activityView{
    if(!_m_activityView){
        _m_activityView=[[ UIActivityIndicatorView alloc]initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
        [self addSubview:_m_activityView];
    }
    return _m_activityView;
}

-(void)funj_showProgressViews:(NSString*)title{
    [self funj_showProgressViews:title t:2 complete:nil];
}
-(void)funj_showProgressViews:(NSString*)title t:(CGFloat)time complete:(completeCallback)complete{
    [self funj_stopProgressAnimate];
    [self.m_timer setFireDate:[NSDate dateWithTimeIntervalSinceNow:time]];
    if(self.m_progressType == kprogressType){
        if(title && title.length>0){
            self.frame = CGRectMake((self.m_superView.width-200)/2, (self.m_superView.height-140)/2, 200, 140);
            self.m_activityView.top = 30;
        }else{
            self.frame = CGRectMake((self.m_superView.width-100)/2, (self.m_superView.height-100)/2, 100, 100);
            self.m_activityView.top = (self.height-self.m_activityView.height)/2;
        }
        [self.m_activityView startAnimating];
        self.m_activityView.left = (self.width-self.m_activityView.width)/2;
    }
    self.m_titleLabel.hidden = YES;
    if(title && title.length>0){
        self.m_titleLabel.hidden = NO;
        [self.m_titleLabel funj_updateAttributedText:title];
        CGFloat top = 0;
        if(self.m_progressType== kprogressType){
            top = _m_activityView.bottom;
        }else{
            CGFloat width =[JAppUtility funj_getTextWidthWithView:self.m_titleLabel]+40;
            if(IS_IPAD){
                width = MIN(KWidth/3*2, width);
            }else{
                width = KWidth>KHeight ? MIN(KWidth/3*2, width) : MIN(KWidth-60,width);
            }
            self.frame = CGRectMake((self.m_superView.width-width)/2, (self.m_superView.height-70)/2, width, 70);
        }
        self.m_titleLabel.frame = CGRectMake(20, top, self.width-40, self.height-top);
    }
    self.m_time = time;
    self.m_blackAlphaView.frame = self.bounds;
    [self.m_superView addSubview:self];
    self.m_isHasFinishCallback = YES;
    self.m_callback = complete;
}
-(void)funj_stopProgressAnimate{
    if(_m_activityView.animating)[_m_activityView stopAnimating];
    if(self.m_timer.isValid)[self.m_timer setFireDate:[NSDate distantFuture]];
    [self removeFromSuperview];
    if(self.m_callback && self.m_isHasFinishCallback){
        self.m_isHasFinishCallback = NO;
        self.m_callback();
    }
    self.m_callback = nil;
}
-(void)dealloc{
    if(self.m_timer.isValid) [self.m_timer invalidate];
}

@end

File: DeveloperArchitecture/DeveloperArchitecture/threeclass/progressHUD/JProgressHUD.swift
//
//  JProgressHUD.swift
//  DeveloperArchitecture
//
//  Created by Jeffrey on 2020/10/30.
//  Copyright © 2020 Jeffery. All rights reserved.
//

import Foundation
import UIKit

protocol JProgressHUDExtApi {
    func funj_showProgressView(_ title : String?)
    
    func funj_stopProgressView()

}

class JProgressHUD: UIView ,JProgressHUDExtApi{
    private var progressBgImageView : UIImageView?
    private var progressImageView : UIImageView?
    
    lazy var m_titleLabel : UILabel = {
        let titleLabel = UILabel(i: CGRect(x: 0, y: 0, width: 0, height: 30), title: nil, textFC: JTextFC(f: kFont_Size17, c: kColor_Text_Black, a: .center))
        self.addSubview(titleLabel)
        return titleLabel
    }()
    override init(frame: CGRect) {
        super.init(frame: CGRect(x: 0, y: 0, width: kWidth, height: kHeight))
        self.backgroundColor = kColor_Clear
        progressBgImageView = UIImageView(i: CGRect(x: (kWidth-202/2.5)/2, y: (kHeight-202/2.5)/2, width: 202/2.5, height: 202/2.5), image: "reloadProgress_center")
        self.addSubview(progressBgImageView!)
        progressImageView = UIImageView(i: CGRect(x: (progressBgImageView!.frame.size.width-188/2.5)/2, y: (progressBgImageView!.frame.size.height-188/2.5)/2, width: 188/2.5, height: 188/2.5), image: "reloadProgress_route")
        progressBgImageView?.addSubview(progressImageView!)
        self.tag = NSNotFound - 1
        let tapGest = UITapGestureRecognizer(target: self, action: #selector(funj_stopProgressView))
        self.addGestureRecognizer(tapGest)
    }
    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented")}
    func funj_addProgressAnimate() {
        let anima = CABasicAnimation(keyPath: "opacity")
        anima.fromValue = 0.3
        anima.toValue = 1.0
        anima.duration = 1.0
        progressBgImageView?.layer.add(anima, forKey: "opacityAniamtion")
        
        let rotationAnimation = CABasicAnimation(keyPath: "transform.rotation.z")
        rotationAnimation.toValue = .pi * 2.0
        rotationAnimation.duration = 0.6
        rotationAnimation.isCumulative = true
        rotationAnimation.repeatCount = HUGE
        progressImageView?.layer.add(rotationAnimation, forKey: "rotationAnimation")
    }
    
    func funj_showProgressView(_ title : String? = nil) {
        let supView = JAppViewTools.funj_getKeyWindow()
        let upProgressView = superview?.viewWithTag(NSNotFound - 1)
        (upProgressView as? JProgressHUD)?.funj_stopProgressView()
        self.frame = CGRect(x: 0,y: 0,width: kWidth,height: kHeight)
        progressBgImageView?.frame = CGRect(x: (kWidth-202/2.5)/2, y: (kHeight-202/2.5)/2, width: 202/2.5, height: 202/2.5)
        
        if title != nil {
            m_titleLabel.frame = CGRect(x: progressBgImageView!.frame.origin.x-20, y: progressBgImageView!.frame.size.height+progressBgImageView!.frame.origin.y, width: progressBgImageView!.frame.size.width+40, height: 30)
            m_titleLabel.text = title
        }
        supView?.addSubview(self)
        self.funj_addProgressAnimate()
    }

    @objc func funj_stopProgressView() {
        progressBgImageView?.layer .removeAllAnimations()
        progressImageView?.layer.removeAllAnimations()
        self.removeFromSuperview()
    }
}
enum MprogressType {
    case kprogressType
    case kprogressType_OnlyText
}
var mprogressHUD : JMProgressHUD? = nil
class JMProgressHUD : UIView {
    var m_blackAlphaView : UIImageView?
    var m_progressType : MprogressType = .kprogressType
    var m_superView : UIView?
    var m_time : TimeInterval = 2
    var m_callback : kcompleteCallback?
    var m_timerIsStart : Bool = false
    
    lazy var m_titleLabel : UILabel = {
        let titleLabel = UILabel(i: CGRectZero, title: nil, textFC: JTextFC(f: kFont_Size15, c: kColor_White, a: .center))
        self.addSubview(titleLabel)
        titleLabel.numberOfLines = 2
        return titleLabel
    }()
    lazy var m_timer : Timer = {
        let timer = Timer.scheduledTimer(timeInterval: 2, target: self, selector: #selector(funj_stopProgressAnimate), userInfo: nil, repeats: true)
        return timer
    }()
    lazy var m_activityView : UIActivityIndicatorView = {
        let activityView = UIActivityIndicatorView(style: .whiteLarge)
        self.addSubview(activityView)
        return activityView
    }()
    
    static let shared: JMProgressHUD = JMProgressHUD()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        funj_addblackView()
    }
    convenience init(superView : UIView , type : MprogressType) {
        self.init(frame : CGRect(x: 0, y: 0, width: 0, height: 0))
    }
    required init?(coder: NSCoder) {  fatalError("init(coder:) has not been implemented")}

    func funj_reloadSuperView(superView : UIView ,type : MprogressType){
        m_progressType = type ; m_superView = superView
    }
    
    func funj_addblackView (){
        m_blackAlphaView = UIImageView(i: self.bounds, bg: kColor_Text_Black)
        .funj_addCornerLayer(JFilletValue(w: 0, r: 10, c: kColor_Clear))
        m_blackAlphaView?.alpha = 0.7
        self.addSubview(m_blackAlphaView!)
    }
    func funj_showProgressViews(title : String? , t time : TimeInterval = 2 ,complete : kcompleteCallback? = nil) {
        funj_stopProgressAnimate()
        m_timerIsStart = true
        m_timer.fireDate = Date.init(timeIntervalSinceNow: time)
        if m_progressType == .kprogressType {
            if title?.count ?? 0 > 0 && m_superView != nil {
                self.frame = CGRect(x: (m_superView!.width - 200 )/2, y: (m_superView!.height - 140 )/2, width: 200, height: 140)
                m_activityView.top = 30;
            } else {
                self.frame = CGRect(x: (m_superView!.width - 100 )/2, y: (m_superView!.height - 100 )/2, width: 100, height: 100)
                m_activityView.top = (self.height - m_activityView.height)/2;
            }
            if kcurrentUserInterfaceStyleModel == 2 /*UIUserInterfaceStyleDark*/ {
                m_activityView.style = .white
            }
            m_activityView.startAnimating()
            m_activityView.left = (self.width - m_activityView.width) / 2
        }
        m_titleLabel.isHidden = true
        if title?.count ?? 0 > 0 {
            m_titleLabel.isHidden = false
            _ = m_titleLabel.funj_updateAttributedText(title!)
            var top : CGFloat = 0.0
            if m_progressType == .kprogressType {
                top = m_activityView.bottom
            } else {
                var width = JAppUtility.funj_getTextWidthWithView(m_titleLabel) + 40
                if kis_IPad {
                    width = min(kWidth / 3 * 2, width)
                } else {
                    width = kWidth > kHeight ? min(kWidth / 3 * 2, width) : min(kWidth - 60 , width)
                }
                self.frame = CGRect(x: ((m_superView?.width ?? 0) - width) / 2, y: ((m_superView?.height ?? 0) - 70 ) / 2, width: width, height: 70)
            }
            m_titleLabel.frame = CGRect(x: 20, y: top, width: self.width - 40 , height: self.height - top)
        }
        m_time = time
        m_blackAlphaView?.frame = self.bounds
        m_superView?.addSubview(self)
        m_callback = complete
    }
    
    @objc private func funj_stopProgressAnimate() {
        if m_activityView.isAnimating { m_activityView.stopAnimating()}
        if m_timerIsStart && m_timer.isValid { m_timer.invalidate()}
        self.removeFromSuperview()
        m_callback?()
        m_callback = nil
    }
    deinit {
        if m_timer.isValid { m_timer.invalidate()}
    }
}
